Support for react-native-source-maps with iOS and Android

reyalpsirc opened the pull request:
- Added support to use react-native-source-maps and give correct JS lines on iOS
- Android also works with the line numbers but they are wrong due to react-native issue https://github.com/facebook/react-native/issues/6946
- Changed the Readme with instructions for both iOS and Android

reyalpsirc committed:
- Added the code to reference react-native-source-maps to get the correct line numbers for JS on iOS.
- Changed the Readme so that it specifies what's needed to have this working.

reyalpsirc committed:
- Updated readme about react-native-fs

mikelambert commented on README.md:
````diff
@@ -4,6 +4,43 @@ react-native-fabric-crashlytics
+You need to install react-native-fs and link it to your main project to run this:
+
+```
+npm install -S react-native-fs
````
Why aren't you listing this as a dependency of this project itself, versus requiring users to install it directly?

Similarly, you should be able to setup this project's rnpm-link to call that project's rnpm-link, right? (So that the instructions would just say to `react-native link react-native-fabric-crashlytics`?)

reyalpsirc replied:
Well, for that I would need to add the iOS and Android folders with the linking parts of react-native-fs itself on the project and I never did that before. Also, I was out of scope already for the project that needed this. Anyway, if someone has experience on that part, it would be better indeed.

mikelambert replied:
Fair :)

mikelambert commented on README.md:
````diff
@@ -4,6 +4,43 @@ react-native-fabric-crashlytics
+react-native link react-native-fs
+```
+
+Setup iOS
````
In theory, these could be done by a rpm-link too, but I get that it becomes more complex, and I'm not as familiar with how to do that, either. :)

At the very least, will your code work if these steps are *not* done?

reyalpsirc replied:
Without react-native-fs linked no. Unless the library is instantiated without the path as parameter, which falls back to the code without sourcemaps. Otherwise, how exactly would this work if there was no way to read the sourcemap?

mikelambert replied:
I think I'm asking "will this still post obfuscated errors if the dev doesn't do this, or is this a breaking change requiring a semver-major bump"?

reyalpsirc replied:
I might get a way around it but the person might be mislead and think that the actual lib is not working.

reyalpsirc replied:
So, the actual question is: should we fallback to the other code if the steps above are not done? Or should a better exception be thrown?

mikelambert replied:
I think I'm okay with making this a requirement, since it is cross-platform if you follow the instructions. I just need to make sure do a major-semver bump after this to avoid anyone accidentally including it.

mikelambert commented on README.md:
```diff
@@ -12,5 +49,9 @@ To use, add this code to your index.ios.js and index.android.js (or some library
\ No newline at end of file
+function crashload () {
+  const path = `sourcemap.js`
+  crashlytics.init(path)
```
I don't understand why you did this function, versus:
`crashlytics.init('sourcemap.js')`

reyalpsirc replied:
Ah, that's because while I was trying to get it working well, I went through so many versions that I forgot to change that back to a simple way

mikelambert commented on package.json:
```diff
@@ -37,7 +37,8 @@
     "lodash.assign": "^4.2.0",
-    "stacktrace-js": "^1.1.0"
+    "stacktrace-js": "^1.3.1",
+    "react-native-source-maps": "git://github.com/philipshurpik/react-native-source-maps.git#master"
```
I've done this in personal projects, but I'm hesitant to stick this into a published npm project (such as this one). Namely, because this project tends to not get updated frequently (I'm not actively using it anymore), and who knows what version-breaking stuff philip might be doing. :)

Is there a particular change you need from his project? Is it easy enough to ask him for a release, versus this change? He seems pretty active on github.

reyalpsirc replied:
The version on npm was not working for me. I don't remember well why but I think it was related to latest versions of reatc native (I'm using 0.49.3)

mikelambert replied:
@philipshurpik (can I tag you?) Any interest in releasing a newer version of your library to npm?

mikelambert replied:
Since I'm not sure if he gets notifications about tags on my repo, can you file an issue on `react-native-source-maps` asking for an npm release?

mikelambert commented on index.js:
```diff
@@ -2,36 +2,68 @@
+        // ignore and use the normal stack on the error
+      }
+      // And then re-throw the exception with the original handler
+      if (originalHandler) {
```
Should we think about unifying these two? Yours is similar-but-different to what I had done:
- We both have delays, though mine was Android-only. (Did you encounter problems on iOS? Or was this motivated by Android problems?)
- Yours seems to handle fatal errors differently, to avoid duplicates, which might be something that could benefit the original one (which screws that aspect up).

reyalpsirc replied:
I did face problems on the iOS but, it might have been Fabric's delay or not. Given that, a small delay there should not cause trouble at all and worked better for me.
When a JS crash happens, it will log a non-fatal AND a fatal error, because it was the only way to get the JS trace and also make the App actually crash on fatals. 
For non-fatals yes, it logs only once.

mikelambert replied:
Makes sense. Do you think you can safely combine your error fallback with the one that existed previously, to avoid duplication?

reyalpsirc replied:
I will try and let you know. I will even try to merge the approaches better so that if there's no stacktrace received from JS, it will use your code.

mikelambert reviewed (commented):
Thanks for this!

reyalpsirc committed:
- Corrected the Readme
- Updated the shell script since it should reflect that its a copy from react-native-xcode.sh and also the changes made
- If there's no JS trace on the sourcemap, the code falls back to the previous code
- The extra fatal error about JS crashes is displayed for both methods so that the user knows that there are JS errors registered as non fatal (since its the only way I found to get a proper trace in time)
- Refactored the code
- Used js standard linter over the code

reyalpsirc commented:
@mikelambert I just improved the code. Please have a look :)

mikelambert commented on index.js:
```diff
@@ -1,43 +1,58 @@
+        Crashlytics.recordCustomExceptionName(e.message, e.message, e.stack)
+        jsParsed = true
+      }
+    } catch (error) {
```
What is this try/catch for? Seems like it would catch any errors in our reporting system...and silently drop them. This seems like a bad idea? Do you know what this is actually catching?

reyalpsirc replied:
The try/catch block was based on the Readme file of https://www.npmjs.com/package/react-native-source-maps. But yeah, I guess it should be only around the line for `await getStackTrace()`. Do you want me to change it? Because I'm currently on vacation and my computer was formatted yesterday and so, I will take a bit more time to look on this again.

mikelambert commented on index.js:
```diff
@@ -1,43 +1,58 @@
-'use strict';
+'use strict'
```
You seem to have rewritten this file without trailing newlines? Was this the result of an auto-formatter or `prettier` run? (If so, we should check in the config). Otherwise, is it possible to keep the newlines for consistency?

reyalpsirc replied:
I have a linter installed on my Atom which uses js-standard. If you want, we can have it back but, I noticed that I was doing stuff like lines without `;` without noticing and then you had them with `;`. That's why I manually formatted the code to the rules of https://standardjs.com/

enthal replied:
@reyalpsirc please respect the coding conventions you see in the files you edit when doing a PR.  Thanks. :) (And wishing your feature was already merged!)

mikelambert reviewed (commented):
Thanks for the cleanup on combining the error-handling!

colinwms commented:
Hey, just wondering what the ETA is for this PR? My team would love to have source map support! For transparency I work at Microsoft.

reyalpsirc committed:
- Corrected the try catch block
- Renamed the sourcemap names to the default ones of iOS and Android

reyalpsirc commented:
@mikelambert I just updated the try catch block and also changed the names of the sourcemaps to their defaults, which made more sense and might help using this together with other libs that use sourcemaps

wootwoot1234 commented:
Merge?  Can't wait to use this.

karthikpala commented:
Hey @reyalpsirc  @mikelambert , this PR is awesome 👍 Can I know why this is not yet merged?
Also some quick questions:
1) Does using react-native-fs add any overhead (or) may be slow down the application as every time it has to use source maps to log error?
2) We are using code-push in our RN app. This would not help us right? As the bundle and source-map will get outdated when a new bundle is created and pushed via code-push?
Any thoughts?

reyalpsirc commented:
Hi @karthikpala 
1. Given the size of the source maps, `fs` will consume a bit of resources to do it, depending on the application complexity.
2. I've not tested `code-push` to much yet but, afaik it allows you to also push the new sourcemap when bundling.

kashif-mujeeb-10p commented:
are we still facing issues for source mapping in android and iOS ?

Actually i integrated the AppCenter SDK but source mapping feature is available there just for iOS only not on Android .

filipef101 committed:
fix bundle map generation

filipef101 committed:
fixed silent crash

enthal commented:
Thanks for this great package!  Any prognosis on getting this merged?  :)

walshie4 commented:
Any news on this PR getting merged?

ronilitman commented:
Is this going to be merged soon?

felixmeziere commented:
Any news on this PR getting merged?
